### DISCUSSION TITLE
python-* packages VS `pip install`ed ones

### DISCUSSION THREAD

**Discussion by stefanwascoding** [Homebrew/discussions#4975] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.25
ORIGIN: https://github.com/Homebrew/brew
HEAD: c32bd1c7cc0c3b1b914845bddfeda53f4d877a3f
Last commit: 5 days ago
Core tap JSON: 12 Dec 11:29 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_AUTO_UPDATE_SECS: 604800
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY_INSTALL_TIMES: set
HOMEBREW_EDITOR: mate -w
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_INSTALL_CLEANUP: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1.2-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  dockutil
  exa
  gcal
  jsonschema
  openssl@1.1
  ufraw

Warning: You have the following deprecated, official taps tapped:
  Homebrew/homebrew-cask-drivers
Untap them with `brew untap`.

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-jinja
  python-idna
  python-click
  pygments
  cffi
  python-attrs
  python-cryptography
```

### Description of issue

I'm using the python installed by homebrew (`/opt/homebrew/opt/python/libexec/bin/python` & `/opt/homebrew/bin/pip`) ... I have had issues upgrading these multiple times with python modules installed as brew dependencies for brew packages (e.g. `python-cryptography`, or `python-idna`) when those same modules are also installed either directly via `pip install idna`, or as a dependency of some python project installed via `pip`.

![python module - brew vs pip](https://github.com/Homebrew/discussions/assets/2292035/3de78b12-8e8d-4b6c-86e1-e1abe5f43c71)

Is there a way to better separate the modules installed via `pip`, and the ones installed via `brew`?
How do I get `brew` & `pip` to play together nicely regarding python modules?

PS: Maybe it's also relevant that the current OS installation was done by upgrades starting from macOS 12.

**Comment by pacman-admin** [comment]
Agreed. Very weird.
They appear not to be interchangeable.
On High Sierra, `pip3 install cryptography` works fine.
However, `brew install python-cryptography` requires tons of weirdo dependencies and stuff.

**Reply by pacman-admin** [reply]
What do you mean by this??
PS: Maybe it's also relevant that the current OS installation was done by upgrades starting from macOS 12.

**Comment by pacman-admin** [comment]
However, sometimes there is no choice.
AFAIK, `pip install tkinter` does not work.
`brew install python-tk` is the only way.

**Comment by pacman-admin** [comment]
The only issue is pip is annoying to upgrade.
I think it was designed primarily for `.venv`s

With brew you just run `brew upgrade`

**Comment by pacman-admin** [comment]
My advice:
use `pip`, it was designed for this.
Only use `brew` for stuff where you want EVERY dependency you could EVER need.
I used git on High Sierra without OPENLDAP and CURL (well, I had the system-installed curl), and it worked perfectly.
Brew is just annoying sometimes.

**Comment by stefanwascoding** [comment]
The problem for me is that some packages/tools require python modules (e.g. `ansible` or `certbot` install `python-cryptography`), and then if you install some other tool via `pip` (maybe because the tool or module isn't available via Homebrew) that _also_ has the `cryptography` module as a requirement, `pip` will happily overwrite the Homebrew module package.  And suddenly you get issues when updating via brew, because it will detect that `pip` installed something in a place where it expected its own files.

**Reply by pacman-admin** [reply]
Brew can be VERY annoying.
Try manually instaling all dependencies that work then run `brew install --ignore-dependencies`

**Reply by stefanwascoding** [reply]
Ignoring the dependencies in the brew install is an interesting idea!  Does it keep ignoring the dependencies during an upgrade as well?

**Comment by pacman-admin** [comment]
Sometimes  weird things happen with`brew upgrade`
I once essentiallybbroke my system by running it.

If anything fails, DO NOT RUN `brew cleanup`
Old versions are kept by default.
Press `Cntl+C`  to abort upgrade if it trys to install things your system hates.
You can run `brew install --ignore-dependencies` to upgrade if needed.

`brew upgrade --ignore-dependencies`  DOES NOT WORK
Try out `brew pin`

**Reply by pacman-admin** [reply]
It generally will continue to ignore dependencies during an upgrade.
